core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -177,76 +177,43 @@ public long sizeInBytes() {
     }
 
     /**
-     * returns a new reader that follows the current writes (most importantly allows making
-     * repeated snapshots that includes new content)
-     */
-    public TranslogReader newReaderFromWriter() {
-        ensureOpen();
-        channelReference.incRef();
-        boolean success = false;
-        try {
-            final TranslogReader reader = new InnerReader(this.generation, firstOperationOffset, channelReference);
-            success = true;
-            return reader;
-        } finally {
-            if (!success) {
-                channelReference.decRef();
-            }
-        }
-    }
-
-    /**
-     * returns a new immutable reader which only exposes the current written operation *
+     * closes this writer and transfers it's underlying file channel to a new immutable reader
      */
-    public ImmutableTranslogReader immutableReader() throws TranslogException {
-        if (channelReference.tryIncRef()) {
-            synchronized (this) {
-                try {
-                    ensureOpen();
-                    outputStream.flush();
-                    ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channelReference, firstOperationOffset, getWrittenOffset(), operationCounter);
-                    channelReference.incRef(); // for new reader
-                    return reader;
-                } catch (Exception e) {
-                    throw new TranslogException(shardId, "exception while creating an immutable reader", e);
-                } finally {
-                    channelReference.decRef();
+    public synchronized ImmutableTranslogReader closeIntoReader() throws TranslogException, IOException {
+        if (closed.compareAndSet(false, true)) {
+            boolean success = false;
+            try {
+                sync();
+                ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
+                success = true;
+                return reader;
+            } catch (Throwable e) {
+                closeWithTragicEvent(e);
+                throw new TranslogException(shardId, "exception while creating an immutable reader", e);
```

Comment:
I really wonder if we should just rethrow and get rid of TranslogException eventually